### DISCUSSION TITLE
How to skip a dependency?

### DISCUSSION THREAD

**Discussion by zoharl3** [microsoft/vcpkg#41788] [Q&A] [answered]
Copying from [here](https://stackoverflow.com/questions/79127253/how-to-skip-a-dependency).

In my manifest, I specify that I want `suitesparse`. It has a dependency `lapack`, and another package (not even sure which) has a dependency `openblas`.

These two dependencies are installed and their libraries are added to my project without making a pip. I don't want them since I use `intel-mkl` instead.

Another scenario, `cgal` installs and sets `boost` without a pip. It decides on a version, and I can't override it.

How do I tell vcpkg to just install a library without its dependencies?

**Comment by autoantwort** [comment]
See here: https://devblogs.microsoft.com/cppblog/using-system-package-manager-dependencies-with-vcpkg/

You create an empty overlay port for these packages

**Reply by zoharl3** [reply]
Thanks, I can see how the overlays can be useful. However, they don't help in both of my cases.

1. To override lapack for suitesparse, I can create an overlay
```
{
    "name": "lapack",
    "version": "0"
}
```
(The version seems meaningless.)
This will skip the installation of lapack. It will also skip the installation of openblas, which is lapack's dependent (doesn't say so on the [website](https://vcpkg.link/ports/lapack)).

This is fine; however, suitesparse still needs blas to built against when vcpkg installs it, and I can't specify at this point to use mkl.

I understand that it's tricky.

2. If I create an empty overlay for boost, its installation is skipped. But I still need it, only a different version. Maybe this should be classified as a bug. I specify in the manifest (no overlay) that I want boost with a specific version. Yet, it's ignored because it's cgal's dependent, which dictates which boost would be installed.

**Reply by autoantwort** [reply]
> This will skip the installation of lapack. It will also skip the installation of openblas, which is lapack's dependent (doesn't say so on the [website](https://vcpkg.link/ports/lapack)).

It says so, it is a transitive dependency (lapack -> clapack -> blas -> openblas)

> suitesparse still needs blas to built against when vcpkg installs it, and I can't specify at this point to use mkl.

You could create an overlay port so that suitesparse uses mkl. 

> If I create an empty overlay for boost, its installation is skipped. But I still need it, only a different version.

Use the versioning feature to specify which version of boost you want: https://learn.microsoft.com/en-us/vcpkg/consume/boost-versions

**Reply by zoharl3** [reply]
Thanks for the first two.

About boost. I've already specified an override version. It works if cgal isn't there. Otherwise, cgal overrides it, and my override is ignored. An example:
```
{
  "dependencies": [
    {
      "name": "cgal"
    },
    {
      "name": "boost"
    }
  ],
  "builtin-baseline": "edc84ff66e6262a9f7565c28eb76577aeab9c5aa",
  "overrides": [
    {
      "name": "cgal",
      "version": "5.6.1"
    },
    {
      "name": "boost",
      "version": "1.80.0"
    }
  ]
}
```
This installs boost 1.86.0 instead of 1.80.0.

**Reply by Osyotr** [reply]
~~I think you should specify boost version constraints in `dependencies` instead of `overrides`.~~
Note: `boost` is a metapackage which includes *all* of boost. You may want to use individual packages instead.

**Reply by autoantwort** [reply, marked as answer]
> About boost. I've already specified an override version. It works if cgal isn't there. Otherwise, cgal overrides it, and my override is ignored. An example:

This is exactly what is explained and solved in https://learn.microsoft.com/en-us/vcpkg/consume/boost-versions
